{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/btsui/anaconda3/envs/deep_nlp_cpu/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n",
      "/cellar/users/btsui/anaconda3/envs/deep_nlp_cpu/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192, got 176\n",
      "  return f(*args, **kwds)\n",
      "/cellar/users/btsui/anaconda3/envs/deep_nlp_cpu/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n",
      "/cellar/users/btsui/anaconda3/envs/deep_nlp_cpu/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192, got 176\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import spacy\n",
    "from spacy.matcher import Matcher\n",
    "import numpy as np\n",
    "#nlp = spacy.load('en_core_web_sm')\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import spacy \n",
    "from tqdm import tqdm\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inS_dir='/cellar/users/btsui/Data/nrnb01_nobackup/METAMAP/allSRS.pickle'\n",
    "\n",
    "sra_dump_pickle_dir='/cellar/users/btsui/Data/SRA/DUMP/sra_dump.pickle'\n",
    "spacy_model_dir='./wikipedia-pubmed-and-PMC-w2v'\n",
    "nlp = spacy.load(spacy_model_dir)\n",
    "srsS=pd.read_pickle(inS_dir)\n",
    "\n",
    "technical_meta_data_df=pd.read_pickle(sra_dump_pickle_dir)\n",
    "\n",
    "tmpDf=technical_meta_data_df.drop_duplicates('Sample')\n",
    "\n",
    "technical_meta_data_df=tmpDf.sample(n=tmpDf.shape[0],random_state=0)\n",
    "\n",
    "#technical_meta_data_df.head()\n",
    "##take only the top ten per study to avoid study bias\n",
    "n=100\n",
    "subsampledSrs=technical_meta_data_df.groupby('Study').head(n=n)['Sample']\n",
    "\n",
    "#subsampledSrs.head()\n",
    "\n",
    "m=srsS.index.get_level_values(0).isin(subsampledSrs.values)\n",
    "\n",
    "srsS_subS=srsS[m]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "srsS_subS.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "attribVC=srsS_subS.index.get_level_values(1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "srs_resetDf=srsS_subS.reset_index()\n",
    "srs_resetDf.columns=['srs','attrib','word']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "recurr_attrib=attribVC.index[attribVC>100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subRecurDf=srs_resetDf[srs_resetDf.attrib.isin(recurr_attrib)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffledDf=subRecurDf.sample(subRecurDf.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsetDf=shuffledDf.groupby('attrib').head(n=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inProcessDf=subsetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "docs=nlp.pipe(inProcessDf['word'].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "vectors=[doc.vector for doc in  tqdm(docs,total=inProcessDf.shape[0])]\n",
    "\n",
    "doc_vectorDf=pd.DataFrame(vectors,index=inProcessDf.set_index(['srs','attrib','word']).index).astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g=doc_vectorDf.groupby('attrib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print ('total # of attributes: ',len(g))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enetityEmbeddingDf=g.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#some of the data didn't have any reads\n",
    "validEmbeddingDf=enetityEmbeddingDf[enetityEmbeddingDf.sum(axis=1).abs()>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enetityEmbeddingDf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def corr_df(df1,df2,metric='cosine' ):\n",
    "    import scipy.spatial as sp\n",
    "    M = 1 - sp.distance.cdist(df1, df2, metric)\n",
    "    print (M.shape)\n",
    "    return pd.DataFrame(data=M,index=df1.index,columns=df2.index)\n",
    "corrDf=corr_df(validEmbeddingDf,validEmbeddingDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corrDf.columns=list(corrDf.index)\n",
    "corrDf.columns.name='attrib2'\n",
    "corrDf.index.name='attrib1'\n",
    "np.fill_diagonal(corrDf.as_matrix(),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corrDf.unstack().dropna().hist(bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "corrDf.loc[:,corrDf.columns.str.contains('scientific')].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#recurr_attribs=attribVC.index[attribVC>100]\n",
    "#organism','geo_loc_name','sex','tissue','genotype','disease'\n",
    "cosine_threshold=0.8\n",
    "#cell/tissue: cell type,  tissue, cell line\n",
    "#,'tissue type'\n",
    "#,'age'\n",
    "\n",
    "['SCIENTIFIC_NAME','genotype','disease','cell type','geo_loc_name','treatment']\n",
    "groupSeedToNewGroupName=pd.Series({'SCIENTIFIC_NAME':'Species',\n",
    "                                  'genotype':'Genotype',\n",
    "                                  'disease':'Disease state',\n",
    "                                  'cell type':'Cell type/tissue',\n",
    "                                   'tissue':'Cell type/tissue',\n",
    "                                  'geo_loc_name':'Geographical Location',\n",
    "                                  'treatment':'Treatment and conditions'})\n",
    "selectedSeedAttribs=groupSeedToNewGroupName.index\n",
    "myDfL=[]\n",
    "for myAttrib in selectedSeedAttribs:\n",
    "    tmpS=corrDf[myAttrib].sort_values(ascending=False)#.loc['source name']#.dropna().sort_values(ascending=False)\n",
    "    tmpDf=tmpS[tmpS>=cosine_threshold].to_frame().reset_index()\n",
    "    tmpDf.columns=['Attributes','cosine']\n",
    "    myDfL.append(tmpDf)\n",
    "    #print(tmpDf[myAttrib])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#geo_loc_name\n",
    "#attribTest='sample_name'\n",
    "\n",
    "\n",
    "#inProcessDf[inProcessDf.attrib==attribTest].head(n=30)#collection_date\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf=pd.concat(myDfL,keys=selectedSeedAttribs,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf.index.names=['GroupName','I']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mergedDf.loc['compound']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf_reset=mergedDf.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf_reset[mergedDf_reset.GroupName=='disease']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#attribVC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf_reset['VC']=attribVC[mergedDf_reset.Attributes].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mergedDf_reset[mergedDf_reset.GroupName\t=='compound']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsetDf=mergedDf_reset[mergedDf_reset['cosine']>=0.8]\n",
    "\n",
    "subsetDf['GroupName']=groupSeedToNewGroupName[subsetDf['GroupName']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsetDf.GroupName.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsetDf.to_csv('./Results/grouping.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!grep -rn grouping.csv *.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 574,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!cat ./Results/grouping.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 559,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GroupName</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cell type/tissue</th>\n",
       "      <td>417924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Disease state</th>\n",
       "      <td>34239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genotype</th>\n",
       "      <td>122909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geographical Location</th>\n",
       "      <td>509997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <td>1416172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treatment and conditions</th>\n",
       "      <td>90516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               VC\n",
       "GroupName                        \n",
       "Cell type/tissue           417924\n",
       "Disease state               34239\n",
       "Genotype                   122909\n",
       "Geographical Location      509997\n",
       "Species                   1416172\n",
       "Treatment and conditions    90516"
      ]
     },
     "execution_count": 559,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsetDf.groupby('GroupName').sum()['VC'].to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GroupName\n",
       "SCIENTIFIC_NAME    0.802767\n",
       "age                0.702257\n",
       "breed              1.000000\n",
       "cell line          0.750326\n",
       "cell type          0.226881\n",
       "dev_stage          0.623596\n",
       "disease            0.632437\n",
       "genotype           0.614813\n",
       "geo_loc_name       0.764516\n",
       "isolate            0.752301\n",
       "sample_type        0.125054\n",
       "sex                0.721619\n",
       "strain             0.920802\n",
       "treatment          0.806940\n",
       "Name: VC, dtype: float64"
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsetDf.groupby('GroupName').first()['VC']/subsetDf.groupby('GroupName').sum()['VC']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GroupName</th>\n",
       "      <th>I</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>cosine</th>\n",
       "      <th>VC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>0</td>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1136856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>1</td>\n",
       "      <td>organism</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>29037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>2</td>\n",
       "      <td>Organism</td>\n",
       "      <td>0.897151</td>\n",
       "      <td>2894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>3</td>\n",
       "      <td>host scientific name</td>\n",
       "      <td>0.863731</td>\n",
       "      <td>9909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>4</td>\n",
       "      <td>Species</td>\n",
       "      <td>0.851593</td>\n",
       "      <td>578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>5</td>\n",
       "      <td>host</td>\n",
       "      <td>0.837771</td>\n",
       "      <td>205511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>6</td>\n",
       "      <td>specific host</td>\n",
       "      <td>0.830889</td>\n",
       "      <td>11114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>7</td>\n",
       "      <td>host_scientific_name</td>\n",
       "      <td>0.826836</td>\n",
       "      <td>4297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>8</td>\n",
       "      <td>host organism</td>\n",
       "      <td>0.825852</td>\n",
       "      <td>372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>9</td>\n",
       "      <td>nat-host</td>\n",
       "      <td>0.819286</td>\n",
       "      <td>1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SCIENTIFIC_NAME</td>\n",
       "      <td>10</td>\n",
       "      <td>specific_host</td>\n",
       "      <td>0.813784</td>\n",
       "      <td>14088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>sex</td>\n",
       "      <td>0</td>\n",
       "      <td>sex</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>163806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>sex</td>\n",
       "      <td>1</td>\n",
       "      <td>host_sex</td>\n",
       "      <td>0.996440</td>\n",
       "      <td>19036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>sex</td>\n",
       "      <td>2</td>\n",
       "      <td>Sex</td>\n",
       "      <td>0.992709</td>\n",
       "      <td>19906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>sex</td>\n",
       "      <td>3</td>\n",
       "      <td>gender</td>\n",
       "      <td>0.989053</td>\n",
       "      <td>18324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>sex</td>\n",
       "      <td>4</td>\n",
       "      <td>host sex</td>\n",
       "      <td>0.987783</td>\n",
       "      <td>3674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>sex</td>\n",
       "      <td>5</td>\n",
       "      <td>host_gender</td>\n",
       "      <td>0.973234</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>sex</td>\n",
       "      <td>6</td>\n",
       "      <td>Gender</td>\n",
       "      <td>0.961932</td>\n",
       "      <td>1030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>sex</td>\n",
       "      <td>7</td>\n",
       "      <td>host-sex</td>\n",
       "      <td>0.938472</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>sex</td>\n",
       "      <td>8</td>\n",
       "      <td>SEX</td>\n",
       "      <td>0.907633</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>sex</td>\n",
       "      <td>9</td>\n",
       "      <td>GENDER</td>\n",
       "      <td>0.803475</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>genotype</td>\n",
       "      <td>0</td>\n",
       "      <td>genotype</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>75566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>genotype</td>\n",
       "      <td>1</td>\n",
       "      <td>genotype/variation</td>\n",
       "      <td>0.971412</td>\n",
       "      <td>28730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>genotype</td>\n",
       "      <td>2</td>\n",
       "      <td>plant genotype</td>\n",
       "      <td>0.877062</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>genotype</td>\n",
       "      <td>3</td>\n",
       "      <td>mutant</td>\n",
       "      <td>0.846111</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>genotype</td>\n",
       "      <td>4</td>\n",
       "      <td>mutation</td>\n",
       "      <td>0.833481</td>\n",
       "      <td>428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>genotype</td>\n",
       "      <td>5</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>0.822699</td>\n",
       "      <td>13892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>genotype</td>\n",
       "      <td>6</td>\n",
       "      <td>host_genotype</td>\n",
       "      <td>0.802237</td>\n",
       "      <td>3784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>disease</td>\n",
       "      <td>0</td>\n",
       "      <td>disease</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>21654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>disease</td>\n",
       "      <td>1</td>\n",
       "      <td>tumor type</td>\n",
       "      <td>0.873200</td>\n",
       "      <td>691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2126</th>\n",
       "      <td>treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>culture condition</td>\n",
       "      <td>0.804804</td>\n",
       "      <td>1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2127</th>\n",
       "      <td>treatment</td>\n",
       "      <td>13</td>\n",
       "      <td>treatment group</td>\n",
       "      <td>0.801314</td>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2128</th>\n",
       "      <td>treatment</td>\n",
       "      <td>14</td>\n",
       "      <td>sample treatment</td>\n",
       "      <td>0.801171</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2747</th>\n",
       "      <td>strain</td>\n",
       "      <td>0</td>\n",
       "      <td>strain</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>273074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2748</th>\n",
       "      <td>strain</td>\n",
       "      <td>1</td>\n",
       "      <td>StrainOrLine</td>\n",
       "      <td>0.899752</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2749</th>\n",
       "      <td>strain</td>\n",
       "      <td>2</td>\n",
       "      <td>genetic background</td>\n",
       "      <td>0.899305</td>\n",
       "      <td>2634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2750</th>\n",
       "      <td>strain</td>\n",
       "      <td>3</td>\n",
       "      <td>host_genotype</td>\n",
       "      <td>0.884557</td>\n",
       "      <td>3784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2751</th>\n",
       "      <td>strain</td>\n",
       "      <td>4</td>\n",
       "      <td>strain/background</td>\n",
       "      <td>0.875381</td>\n",
       "      <td>3933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2752</th>\n",
       "      <td>strain</td>\n",
       "      <td>5</td>\n",
       "      <td>strain or line</td>\n",
       "      <td>0.873404</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2753</th>\n",
       "      <td>strain</td>\n",
       "      <td>6</td>\n",
       "      <td>strain background</td>\n",
       "      <td>0.871195</td>\n",
       "      <td>6334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2754</th>\n",
       "      <td>strain</td>\n",
       "      <td>7</td>\n",
       "      <td>strain name</td>\n",
       "      <td>0.867428</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2755</th>\n",
       "      <td>strain</td>\n",
       "      <td>8</td>\n",
       "      <td>background strain</td>\n",
       "      <td>0.866178</td>\n",
       "      <td>2099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2756</th>\n",
       "      <td>strain</td>\n",
       "      <td>9</td>\n",
       "      <td>strain/genotype</td>\n",
       "      <td>0.846380</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2757</th>\n",
       "      <td>strain</td>\n",
       "      <td>10</td>\n",
       "      <td>host genotype</td>\n",
       "      <td>0.843881</td>\n",
       "      <td>720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2758</th>\n",
       "      <td>strain</td>\n",
       "      <td>11</td>\n",
       "      <td>stain</td>\n",
       "      <td>0.838131</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2759</th>\n",
       "      <td>strain</td>\n",
       "      <td>12</td>\n",
       "      <td>host strain</td>\n",
       "      <td>0.833982</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2760</th>\n",
       "      <td>strain</td>\n",
       "      <td>13</td>\n",
       "      <td>mouse strain</td>\n",
       "      <td>0.831673</td>\n",
       "      <td>604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2761</th>\n",
       "      <td>strain</td>\n",
       "      <td>14</td>\n",
       "      <td>host_strain</td>\n",
       "      <td>0.830388</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2762</th>\n",
       "      <td>strain</td>\n",
       "      <td>15</td>\n",
       "      <td>paternal strain</td>\n",
       "      <td>0.826031</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2763</th>\n",
       "      <td>strain</td>\n",
       "      <td>16</td>\n",
       "      <td>strain description</td>\n",
       "      <td>0.812128</td>\n",
       "      <td>620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2764</th>\n",
       "      <td>strain</td>\n",
       "      <td>17</td>\n",
       "      <td>background</td>\n",
       "      <td>0.809500</td>\n",
       "      <td>604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2765</th>\n",
       "      <td>strain</td>\n",
       "      <td>18</td>\n",
       "      <td>host infra-specific name</td>\n",
       "      <td>0.805565</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2924</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>0</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>79503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2925</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>1</td>\n",
       "      <td>sample type</td>\n",
       "      <td>0.902240</td>\n",
       "      <td>10466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2926</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>2</td>\n",
       "      <td>description</td>\n",
       "      <td>0.861178</td>\n",
       "      <td>19639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2927</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>3</td>\n",
       "      <td>DESCRIPTION</td>\n",
       "      <td>0.847247</td>\n",
       "      <td>310472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2928</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>4</td>\n",
       "      <td>sample-type</td>\n",
       "      <td>0.846547</td>\n",
       "      <td>582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2929</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>5</td>\n",
       "      <td>host_tissue_sampled</td>\n",
       "      <td>0.806701</td>\n",
       "      <td>13772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2930</th>\n",
       "      <td>sample_type</td>\n",
       "      <td>6</td>\n",
       "      <td>isolation_source</td>\n",
       "      <td>0.803889</td>\n",
       "      <td>201316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3228</th>\n",
       "      <td>breed</td>\n",
       "      <td>0</td>\n",
       "      <td>breed</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>44001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>228 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GroupName   I                Attributes    cosine       VC\n",
       "0     SCIENTIFIC_NAME   0           SCIENTIFIC_NAME  1.000000  1136856\n",
       "1     SCIENTIFIC_NAME   1                  organism  0.983355    29037\n",
       "2     SCIENTIFIC_NAME   2                  Organism  0.897151     2894\n",
       "3     SCIENTIFIC_NAME   3      host scientific name  0.863731     9909\n",
       "4     SCIENTIFIC_NAME   4                   Species  0.851593      578\n",
       "5     SCIENTIFIC_NAME   5                      host  0.837771   205511\n",
       "6     SCIENTIFIC_NAME   6             specific host  0.830889    11114\n",
       "7     SCIENTIFIC_NAME   7      host_scientific_name  0.826836     4297\n",
       "8     SCIENTIFIC_NAME   8             host organism  0.825852      372\n",
       "9     SCIENTIFIC_NAME   9                  nat-host  0.819286     1516\n",
       "10    SCIENTIFIC_NAME  10             specific_host  0.813784    14088\n",
       "50                sex   0                       sex  1.000000   163806\n",
       "51                sex   1                  host_sex  0.996440    19036\n",
       "52                sex   2                       Sex  0.992709    19906\n",
       "53                sex   3                    gender  0.989053    18324\n",
       "54                sex   4                  host sex  0.987783     3674\n",
       "55                sex   5               host_gender  0.973234      313\n",
       "56                sex   6                    Gender  0.961932     1030\n",
       "57                sex   7                  host-sex  0.938472      207\n",
       "58                sex   8                       SEX  0.907633      289\n",
       "59                sex   9                    GENDER  0.803475      413\n",
       "93           genotype   0                  genotype  1.000000    75566\n",
       "94           genotype   1        genotype/variation  0.971412    28730\n",
       "95           genotype   2            plant genotype  0.877062      195\n",
       "96           genotype   3                    mutant  0.846111      314\n",
       "97           genotype   4                  mutation  0.833481      428\n",
       "98           genotype   5                 phenotype  0.822699    13892\n",
       "99           genotype   6             host_genotype  0.802237     3784\n",
       "374           disease   0                   disease  1.000000    21654\n",
       "375           disease   1                tumor type  0.873200      691\n",
       "...               ...  ..                       ...       ...      ...\n",
       "2126        treatment  12         culture condition  0.804804     1220\n",
       "2127        treatment  13           treatment group  0.801314      405\n",
       "2128        treatment  14          sample treatment  0.801171      102\n",
       "2747           strain   0                    strain  1.000000   273074\n",
       "2748           strain   1              StrainOrLine  0.899752      660\n",
       "2749           strain   2        genetic background  0.899305     2634\n",
       "2750           strain   3             host_genotype  0.884557     3784\n",
       "2751           strain   4         strain/background  0.875381     3933\n",
       "2752           strain   5            strain or line  0.873404      169\n",
       "2753           strain   6         strain background  0.871195     6334\n",
       "2754           strain   7               strain name  0.867428      104\n",
       "2755           strain   8         background strain  0.866178     2099\n",
       "2756           strain   9           strain/genotype  0.846380      235\n",
       "2757           strain  10             host genotype  0.843881      720\n",
       "2758           strain  11                     stain  0.838131      207\n",
       "2759           strain  12               host strain  0.833982      250\n",
       "2760           strain  13              mouse strain  0.831673      604\n",
       "2761           strain  14               host_strain  0.830388      154\n",
       "2762           strain  15           paternal strain  0.826031      162\n",
       "2763           strain  16        strain description  0.812128      620\n",
       "2764           strain  17                background  0.809500      604\n",
       "2765           strain  18  host infra-specific name  0.805565      214\n",
       "2924      sample_type   0               sample_type  1.000000    79503\n",
       "2925      sample_type   1               sample type  0.902240    10466\n",
       "2926      sample_type   2               description  0.861178    19639\n",
       "2927      sample_type   3               DESCRIPTION  0.847247   310472\n",
       "2928      sample_type   4               sample-type  0.846547      582\n",
       "2929      sample_type   5       host_tissue_sampled  0.806701    13772\n",
       "2930      sample_type   6          isolation_source  0.803889   201316\n",
       "3228            breed   0                     breed  1.000000    44001\n",
       "\n",
       "[228 rows x 5 columns]"
      ]
     },
     "execution_count": 400,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedDf_reset[mergedDf_reset['cosine']>0.8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {},
   "outputs": [],
   "source": [
    "recurr_attribs=attribVC.index[attribVC>100]\n",
    "cosinDf=corrDf[attribTest].sort_values(ascending=False).head(n=20).to_frame()#.loc['source name']#.dropna().sort_values(ascending=False)\n",
    "cosinDf['n']=attribVC.loc[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>attrib1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>30517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_title</th>\n",
       "      <td>0.802975</td>\n",
       "      <td>22382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUBMITTER_ID</th>\n",
       "      <td>0.787158</td>\n",
       "      <td>747357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_material_id</th>\n",
       "      <td>0.783863</td>\n",
       "      <td>23510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <td>0.781657</td>\n",
       "      <td>19838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_subject_id</th>\n",
       "      <td>0.773550</td>\n",
       "      <td>36272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_ID</th>\n",
       "      <td>0.773218</td>\n",
       "      <td>2240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample id</th>\n",
       "      <td>0.772147</td>\n",
       "      <td>1095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>code</th>\n",
       "      <td>0.763271</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient id</th>\n",
       "      <td>0.743997</td>\n",
       "      <td>1518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample ID</th>\n",
       "      <td>0.742087</td>\n",
       "      <td>986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submitted sample id</th>\n",
       "      <td>0.739071</td>\n",
       "      <td>31454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isolate_name_alias</th>\n",
       "      <td>0.738362</td>\n",
       "      <td>5816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Title</th>\n",
       "      <td>0.735538</td>\n",
       "      <td>735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>well id</th>\n",
       "      <td>0.734158</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <td>0.730197</td>\n",
       "      <td>2399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submitted subject id</th>\n",
       "      <td>0.724277</td>\n",
       "      <td>31450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>colony</th>\n",
       "      <td>0.723876</td>\n",
       "      <td>813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>well_location</th>\n",
       "      <td>0.722963</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>local_family_id</th>\n",
       "      <td>0.716479</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      sample_name       n\n",
       "attrib1                                  \n",
       "sample_name              1.000000   30517\n",
       "sample_title             0.802975   22382\n",
       "SUBMITTER_ID             0.787158  747357\n",
       "source_material_id       0.783863   23510\n",
       "label                    0.781657   19838\n",
       "host_subject_id          0.773550   36272\n",
       "sample_ID                0.773218    2240\n",
       "sample id                0.772147    1095\n",
       "code                     0.763271     721\n",
       "patient id               0.743997    1518\n",
       "Sample ID                0.742087     986\n",
       "submitted sample id      0.739071   31454\n",
       "isolate_name_alias       0.738362    5816\n",
       "Title                    0.735538     735\n",
       "well id                  0.734158     200\n",
       "sampleID                 0.730197    2399\n",
       "submitted subject id     0.724277   31450\n",
       "colony                   0.723876     813\n",
       "well_location            0.722963     147\n",
       "local_family_id          0.716479     829"
      ]
     },
     "execution_count": 402,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosinDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print (\"\\n\".join(attribVC.head(n=100).to_frame().index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nhttps://metamap.nlm.nih.gov/Docs/SemGroups_2013.txt\\n\\nActivities & Behaviors\\nAnatomy\\nChemicals & Drugs\\nConcepts & Ideas\\nDevices\\nDisorders\\nGenes & Molecular Sequences\\nLiving Beings\\nGeographic Areas\\nObjects\\nOrganizations\\nPhysiology\\nProcedures\\n\\nACTI|Activities & Behaviors|T052|Activity\\nACTI|Activities & Behaviors|T053|Behavior\\nACTI|Activities & Behaviors|T056|Daily or Recreational Activity\\nACTI|Activities & Behaviors|T051|Event\\nACTI|Activities & Behaviors|T064|Governmental or Regulatory Activity\\nACTI|Activities & Behaviors|T055|Individual Behavior\\nACTI|Activities & Behaviors|T066|Machine Activity\\nACTI|Activities & Behaviors|T057|Occupational Activity\\nACTI|Activities & Behaviors|T054|Social Behavior\\nANAT|Anatomy|T017|Anatomical Structure\\nANAT|Anatomy|T029|Body Location or Region\\nANAT|Anatomy|T023|Body Part, Organ, or Organ Component\\nANAT|Anatomy|T030|Body Space or Junction\\nANAT|Anatomy|T031|Body Substance\\nANAT|Anatomy|T022|Body System\\nANAT|Anatomy|T025|Cell\\nANAT|Anatomy|T026|Cell Component\\nANAT|Anatomy|T018|Embryonic Structure\\nANAT|Anatomy|T021|Fully Formed Anatomical Structure\\nANAT|Anatomy|T024|Tissue\\nCHEM|Chemicals & Drugs|T116|Amino Acid, Peptide, or Protein\\nCHEM|Chemicals & Drugs|T195|Antibiotic\\nCHEM|Chemicals & Drugs|T123|Biologically Active Substance\\nCHEM|Chemicals & Drugs|T122|Biomedical or Dental Material\\nCHEM|Chemicals & Drugs|T118|Carbohydrate\\nCHEM|Chemicals & Drugs|T103|Chemical\\nCHEM|Chemicals & Drugs|T120|Chemical Viewed Functionally\\nCHEM|Chemicals & Drugs|T104|Chemical Viewed Structurally\\nCHEM|Chemicals & Drugs|T200|Clinical Drug\\nCHEM|Chemicals & Drugs|T111|Eicosanoid\\nCHEM|Chemicals & Drugs|T196|Element, Ion, or Isotope\\nCHEM|Chemicals & Drugs|T126|Enzyme\\nCHEM|Chemicals & Drugs|T131|Hazardous or Poisonous Substance\\nCHEM|Chemicals & Drugs|T125|Hormone\\nCHEM|Chemicals & Drugs|T129|Immunologic Factor\\nCHEM|Chemicals & Drugs|T130|Indicator, Reagent, or Diagnostic Aid\\nCHEM|Chemicals & Drugs|T197|Inorganic Chemical\\nCHEM|Chemicals & Drugs|T119|Lipid\\nCHEM|Chemicals & Drugs|T124|Neuroreactive Substance or Biogenic Amine\\nCHEM|Chemicals & Drugs|T114|Nucleic Acid, Nucleoside, or Nucleotide\\nCHEM|Chemicals & Drugs|T109|Organic Chemical\\nCHEM|Chemicals & Drugs|T115|Organophosphorus Compound\\nCHEM|Chemicals & Drugs|T121|Pharmacologic Substance\\nCHEM|Chemicals & Drugs|T192|Receptor\\nCHEM|Chemicals & Drugs|T110|Steroid\\nCHEM|Chemicals & Drugs|T127|Vitamin\\nCONC|Concepts & Ideas|T185|Classification\\nCONC|Concepts & Ideas|T077|Conceptual Entity\\nCONC|Concepts & Ideas|T169|Functional Concept\\nCONC|Concepts & Ideas|T102|Group Attribute\\nCONC|Concepts & Ideas|T078|Idea or Concept\\nCONC|Concepts & Ideas|T170|Intellectual Product\\nCONC|Concepts & Ideas|T171|Language\\nCONC|Concepts & Ideas|T080|Qualitative Concept\\nCONC|Concepts & Ideas|T081|Quantitative Concept\\nCONC|Concepts & Ideas|T089|Regulation or Law\\nCONC|Concepts & Ideas|T082|Spatial Concept\\nCONC|Concepts & Ideas|T079|Temporal Concept\\nDEVI|Devices|T203|Drug Delivery Device\\nDEVI|Devices|T074|Medical Device\\nDEVI|Devices|T075|Research Device\\nDISO|Disorders|T020|Acquired Abnormality\\nDISO|Disorders|T190|Anatomical Abnormality\\nDISO|Disorders|T049|Cell or Molecular Dysfunction\\nDISO|Disorders|T019|Congenital Abnormality\\nDISO|Disorders|T047|Disease or Syndrome\\nDISO|Disorders|T050|Experimental Model of Disease\\nDISO|Disorders|T033|Finding\\nDISO|Disorders|T037|Injury or Poisoning\\nDISO|Disorders|T048|Mental or Behavioral Dysfunction\\nDISO|Disorders|T191|Neoplastic Process\\nDISO|Disorders|T046|Pathologic Function\\nDISO|Disorders|T184|Sign or Symptom\\nGENE|Genes & Molecular Sequences|T087|Amino Acid Sequence\\nGENE|Genes & Molecular Sequences|T088|Carbohydrate Sequence\\nGENE|Genes & Molecular Sequences|T028|Gene or Genome\\nGENE|Genes & Molecular Sequences|T085|Molecular Sequence\\nGENE|Genes & Molecular Sequences|T086|Nucleotide Sequence\\nGEOG|Geographic Areas|T083|Geographic Area\\nLIVB|Living Beings|T100|Age Group\\nLIVB|Living Beings|T011|Amphibian\\nLIVB|Living Beings|T008|Animal\\nLIVB|Living Beings|T194|Archaeon\\nLIVB|Living Beings|T007|Bacterium\\nLIVB|Living Beings|T012|Bird\\nLIVB|Living Beings|T204|Eukaryote\\nLIVB|Living Beings|T099|Family Group\\nLIVB|Living Beings|T013|Fish\\nLIVB|Living Beings|T004|Fungus\\nLIVB|Living Beings|T096|Group\\nLIVB|Living Beings|T016|Human\\nLIVB|Living Beings|T015|Mammal\\nLIVB|Living Beings|T001|Organism\\nLIVB|Living Beings|T101|Patient or Disabled Group\\nLIVB|Living Beings|T002|Plant\\nLIVB|Living Beings|T098|Population Group\\nLIVB|Living Beings|T097|Professional or Occupational Group\\nLIVB|Living Beings|T014|Reptile\\nLIVB|Living Beings|T010|Vertebrate\\nLIVB|Living Beings|T005|Virus\\nOBJC|Objects|T071|Entity\\nOBJC|Objects|T168|Food\\nOBJC|Objects|T073|Manufactured Object\\nOBJC|Objects|T072|Physical Object\\nOBJC|Objects|T167|Substance\\nOCCU|Occupations|T091|Biomedical Occupation or Discipline\\nOCCU|Occupations|T090|Occupation or Discipline\\nORGA|Organizations|T093|Health Care Related Organization\\nORGA|Organizations|T092|Organization\\nORGA|Organizations|T094|Professional Society\\nORGA|Organizations|T095|Self-help or Relief Organization\\nPHEN|Phenomena|T038|Biologic Function\\nPHEN|Phenomena|T069|Environmental Effect of Humans\\nPHEN|Phenomena|T068|Human-caused Phenomenon or Process\\nPHEN|Phenomena|T034|Laboratory or Test Result\\nPHEN|Phenomena|T070|Natural Phenomenon or Process\\nPHEN|Phenomena|T067|Phenomenon or Process\\nPHYS|Physiology|T043|Cell Function\\nPHYS|Physiology|T201|Clinical Attribute\\nPHYS|Physiology|T045|Genetic Function\\nPHYS|Physiology|T041|Mental Process\\nPHYS|Physiology|T044|Molecular Function\\nPHYS|Physiology|T032|Organism Attribute\\nPHYS|Physiology|T040|Organism Function\\nPHYS|Physiology|T042|Organ or Tissue Function\\nPHYS|Physiology|T039|Physiologic Function\\nPROC|Procedures|T060|Diagnostic Procedure\\nPROC|Procedures|T065|Educational Activity\\nPROC|Procedures|T058|Health Care Activity\\nPROC|Procedures|T059|Laboratory Procedure\\nPROC|Procedures|T063|Molecular Biology Research Technique\\nPROC|Procedures|T062|Research Activity\\nPROC|Procedures|T061|Therapeutic or Preventive Procedure\\n'"
      ]
     },
     "execution_count": 404,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "https://metamap.nlm.nih.gov/Docs/SemGroups_2013.txt\n",
    "\n",
    "Activities & Behaviors\n",
    "Anatomy\n",
    "Chemicals & Drugs\n",
    "Concepts & Ideas\n",
    "Devices\n",
    "Disorders\n",
    "Genes & Molecular Sequences\n",
    "Living Beings\n",
    "Geographic Areas\n",
    "Objects\n",
    "Organizations\n",
    "Physiology\n",
    "Procedures\n",
    "\n",
    "ACTI|Activities & Behaviors|T052|Activity\n",
    "ACTI|Activities & Behaviors|T053|Behavior\n",
    "ACTI|Activities & Behaviors|T056|Daily or Recreational Activity\n",
    "ACTI|Activities & Behaviors|T051|Event\n",
    "ACTI|Activities & Behaviors|T064|Governmental or Regulatory Activity\n",
    "ACTI|Activities & Behaviors|T055|Individual Behavior\n",
    "ACTI|Activities & Behaviors|T066|Machine Activity\n",
    "ACTI|Activities & Behaviors|T057|Occupational Activity\n",
    "ACTI|Activities & Behaviors|T054|Social Behavior\n",
    "ANAT|Anatomy|T017|Anatomical Structure\n",
    "ANAT|Anatomy|T029|Body Location or Region\n",
    "ANAT|Anatomy|T023|Body Part, Organ, or Organ Component\n",
    "ANAT|Anatomy|T030|Body Space or Junction\n",
    "ANAT|Anatomy|T031|Body Substance\n",
    "ANAT|Anatomy|T022|Body System\n",
    "ANAT|Anatomy|T025|Cell\n",
    "ANAT|Anatomy|T026|Cell Component\n",
    "ANAT|Anatomy|T018|Embryonic Structure\n",
    "ANAT|Anatomy|T021|Fully Formed Anatomical Structure\n",
    "ANAT|Anatomy|T024|Tissue\n",
    "CHEM|Chemicals & Drugs|T116|Amino Acid, Peptide, or Protein\n",
    "CHEM|Chemicals & Drugs|T195|Antibiotic\n",
    "CHEM|Chemicals & Drugs|T123|Biologically Active Substance\n",
    "CHEM|Chemicals & Drugs|T122|Biomedical or Dental Material\n",
    "CHEM|Chemicals & Drugs|T118|Carbohydrate\n",
    "CHEM|Chemicals & Drugs|T103|Chemical\n",
    "CHEM|Chemicals & Drugs|T120|Chemical Viewed Functionally\n",
    "CHEM|Chemicals & Drugs|T104|Chemical Viewed Structurally\n",
    "CHEM|Chemicals & Drugs|T200|Clinical Drug\n",
    "CHEM|Chemicals & Drugs|T111|Eicosanoid\n",
    "CHEM|Chemicals & Drugs|T196|Element, Ion, or Isotope\n",
    "CHEM|Chemicals & Drugs|T126|Enzyme\n",
    "CHEM|Chemicals & Drugs|T131|Hazardous or Poisonous Substance\n",
    "CHEM|Chemicals & Drugs|T125|Hormone\n",
    "CHEM|Chemicals & Drugs|T129|Immunologic Factor\n",
    "CHEM|Chemicals & Drugs|T130|Indicator, Reagent, or Diagnostic Aid\n",
    "CHEM|Chemicals & Drugs|T197|Inorganic Chemical\n",
    "CHEM|Chemicals & Drugs|T119|Lipid\n",
    "CHEM|Chemicals & Drugs|T124|Neuroreactive Substance or Biogenic Amine\n",
    "CHEM|Chemicals & Drugs|T114|Nucleic Acid, Nucleoside, or Nucleotide\n",
    "CHEM|Chemicals & Drugs|T109|Organic Chemical\n",
    "CHEM|Chemicals & Drugs|T115|Organophosphorus Compound\n",
    "CHEM|Chemicals & Drugs|T121|Pharmacologic Substance\n",
    "CHEM|Chemicals & Drugs|T192|Receptor\n",
    "CHEM|Chemicals & Drugs|T110|Steroid\n",
    "CHEM|Chemicals & Drugs|T127|Vitamin\n",
    "CONC|Concepts & Ideas|T185|Classification\n",
    "CONC|Concepts & Ideas|T077|Conceptual Entity\n",
    "CONC|Concepts & Ideas|T169|Functional Concept\n",
    "CONC|Concepts & Ideas|T102|Group Attribute\n",
    "CONC|Concepts & Ideas|T078|Idea or Concept\n",
    "CONC|Concepts & Ideas|T170|Intellectual Product\n",
    "CONC|Concepts & Ideas|T171|Language\n",
    "CONC|Concepts & Ideas|T080|Qualitative Concept\n",
    "CONC|Concepts & Ideas|T081|Quantitative Concept\n",
    "CONC|Concepts & Ideas|T089|Regulation or Law\n",
    "CONC|Concepts & Ideas|T082|Spatial Concept\n",
    "CONC|Concepts & Ideas|T079|Temporal Concept\n",
    "DEVI|Devices|T203|Drug Delivery Device\n",
    "DEVI|Devices|T074|Medical Device\n",
    "DEVI|Devices|T075|Research Device\n",
    "DISO|Disorders|T020|Acquired Abnormality\n",
    "DISO|Disorders|T190|Anatomical Abnormality\n",
    "DISO|Disorders|T049|Cell or Molecular Dysfunction\n",
    "DISO|Disorders|T019|Congenital Abnormality\n",
    "DISO|Disorders|T047|Disease or Syndrome\n",
    "DISO|Disorders|T050|Experimental Model of Disease\n",
    "DISO|Disorders|T033|Finding\n",
    "DISO|Disorders|T037|Injury or Poisoning\n",
    "DISO|Disorders|T048|Mental or Behavioral Dysfunction\n",
    "DISO|Disorders|T191|Neoplastic Process\n",
    "DISO|Disorders|T046|Pathologic Function\n",
    "DISO|Disorders|T184|Sign or Symptom\n",
    "GENE|Genes & Molecular Sequences|T087|Amino Acid Sequence\n",
    "GENE|Genes & Molecular Sequences|T088|Carbohydrate Sequence\n",
    "GENE|Genes & Molecular Sequences|T028|Gene or Genome\n",
    "GENE|Genes & Molecular Sequences|T085|Molecular Sequence\n",
    "GENE|Genes & Molecular Sequences|T086|Nucleotide Sequence\n",
    "GEOG|Geographic Areas|T083|Geographic Area\n",
    "LIVB|Living Beings|T100|Age Group\n",
    "LIVB|Living Beings|T011|Amphibian\n",
    "LIVB|Living Beings|T008|Animal\n",
    "LIVB|Living Beings|T194|Archaeon\n",
    "LIVB|Living Beings|T007|Bacterium\n",
    "LIVB|Living Beings|T012|Bird\n",
    "LIVB|Living Beings|T204|Eukaryote\n",
    "LIVB|Living Beings|T099|Family Group\n",
    "LIVB|Living Beings|T013|Fish\n",
    "LIVB|Living Beings|T004|Fungus\n",
    "LIVB|Living Beings|T096|Group\n",
    "LIVB|Living Beings|T016|Human\n",
    "LIVB|Living Beings|T015|Mammal\n",
    "LIVB|Living Beings|T001|Organism\n",
    "LIVB|Living Beings|T101|Patient or Disabled Group\n",
    "LIVB|Living Beings|T002|Plant\n",
    "LIVB|Living Beings|T098|Population Group\n",
    "LIVB|Living Beings|T097|Professional or Occupational Group\n",
    "LIVB|Living Beings|T014|Reptile\n",
    "LIVB|Living Beings|T010|Vertebrate\n",
    "LIVB|Living Beings|T005|Virus\n",
    "OBJC|Objects|T071|Entity\n",
    "OBJC|Objects|T168|Food\n",
    "OBJC|Objects|T073|Manufactured Object\n",
    "OBJC|Objects|T072|Physical Object\n",
    "OBJC|Objects|T167|Substance\n",
    "OCCU|Occupations|T091|Biomedical Occupation or Discipline\n",
    "OCCU|Occupations|T090|Occupation or Discipline\n",
    "ORGA|Organizations|T093|Health Care Related Organization\n",
    "ORGA|Organizations|T092|Organization\n",
    "ORGA|Organizations|T094|Professional Society\n",
    "ORGA|Organizations|T095|Self-help or Relief Organization\n",
    "PHEN|Phenomena|T038|Biologic Function\n",
    "PHEN|Phenomena|T069|Environmental Effect of Humans\n",
    "PHEN|Phenomena|T068|Human-caused Phenomenon or Process\n",
    "PHEN|Phenomena|T034|Laboratory or Test Result\n",
    "PHEN|Phenomena|T070|Natural Phenomenon or Process\n",
    "PHEN|Phenomena|T067|Phenomenon or Process\n",
    "PHYS|Physiology|T043|Cell Function\n",
    "PHYS|Physiology|T201|Clinical Attribute\n",
    "PHYS|Physiology|T045|Genetic Function\n",
    "PHYS|Physiology|T041|Mental Process\n",
    "PHYS|Physiology|T044|Molecular Function\n",
    "PHYS|Physiology|T032|Organism Attribute\n",
    "PHYS|Physiology|T040|Organism Function\n",
    "PHYS|Physiology|T042|Organ or Tissue Function\n",
    "PHYS|Physiology|T039|Physiologic Function\n",
    "PROC|Procedures|T060|Diagnostic Procedure\n",
    "PROC|Procedures|T065|Educational Activity\n",
    "PROC|Procedures|T058|Health Care Activity\n",
    "PROC|Procedures|T059|Laboratory Procedure\n",
    "PROC|Procedures|T063|Molecular Biology Research Technique\n",
    "PROC|Procedures|T062|Research Activity\n",
    "PROC|Procedures|T061|Therapeutic or Preventive Procedure\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "m=srsS_subS.str.contains('fac',case=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [],
   "source": [
    "VC=m.astype(int).groupby(level=1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "missing                                                                                                15801\n",
       "not applicable                                                                                         11323\n",
       "soil                                                                                                    4472\n",
       "feces                                                                                                   3779\n",
       "not collected                                                                                           2625\n",
       "stool                                                                                                   2572\n",
       "Blood                                                                                                   1911\n",
       "sputum                                                                                                  1783\n",
       "Human                                                                                                   1754\n",
       "water                                                                                                   1527\n",
       "Feces                                                                                                   1482\n",
       "blood                                                                                                   1450\n",
       "NA                                                                                                      1218\n",
       "Sputum                                                                                                  1189\n",
       "not known                                                                                               1128\n",
       "Soil                                                                                                    1022\n",
       "nasopharynx                                                                                             1022\n",
       "seawater                                                                                                1001\n",
       "faeces                                                                                                   964\n",
       "Unknown                                                                                                  957\n",
       "N/A                                                                                                      954\n",
       "Not applicable                                                                                           942\n",
       "sediment                                                                                                 911\n",
       "Faeces                                                                                                   891\n",
       "skin                                                                                                     837\n",
       "laboratory                                                                                               819\n",
       "Stool                                                                                                    806\n",
       "soil sample                                                                                              749\n",
       "Missing                                                                                                  733\n",
       "urine                                                                                                    688\n",
       "                                                                                                       ...  \n",
       "SA_D9,D9                                                                                                   1\n",
       "Spawn Microcosm 66 h_2                                                                                     1\n",
       "buccal cavity of Sprague Dawley rats 4 hours  post mortem in the sarcosaphagous insect access group        1\n",
       "Municipal wastewater and rhizospheric soil                                                                 1\n",
       "Escherichia_coli                                                                                           1\n",
       "Paniphala hot spring flowing water                                                                         1\n",
       "Prunus orchards122                                                                                         1\n",
       "Wastewater_November_T0_Control                                                                             1\n",
       "sludge sample of R1                                                                                        1\n",
       "bronchoalveolar lavage fluid from cystic fibrosis patient at Johns Hopkins Hospital                        1\n",
       "cin66                                                                                                      1\n",
       "Clinical isolate; obtained from J. Bakken                                                                  1\n",
       "Faecal of H14                                                                                              1\n",
       "Control (seed) Run 3 end DNA vessel 1                                                                      1\n",
       "sand from beach, distant from spring and 3m vertical displacement                                          1\n",
       "W0d (ii)                                                                                                   1\n",
       "clone H01                                                                                                  1\n",
       "6_20                                                                                                       1\n",
       "raw meat_M_57                                                                                              1\n",
       "regenerating livers collected at 24h3                                                                      1\n",
       "Whole_population Batch_18-Rep_3-t_1                                                                        1\n",
       "GALV-4-88 GALV-Brain infected bat lung fibroblasts                                                         1\n",
       "fish (foodborne isolate)                                                                                   1\n",
       "BIGSid709_Requeijao (curd cheese)                                                                          1\n",
       "Secondary effluent after reactivation-WWTP-Q2                                                              1\n",
       "SSU rRNA amplicon from RNA, Top core section from HL9                                                      1\n",
       "Evolution Experiment Location 200                                                                          1\n",
       "brackish sediment, B5 48 h                                                                                 1\n",
       "LMO-NB1-LMO_Replicate b                                                                                    1\n",
       "Jackson mice gavaged with JAX 7d                                                                           1\n",
       "Length: 25439, dtype: int64"
      ]
     },
     "execution_count": 369,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srsS_subS[srsS_subS.index.get_level_values(1)=='isolation_source'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Isolation Site                    23\n",
       "separation method                 23\n",
       "study disease                     24\n",
       "facs sorting                      24\n",
       "anonymized name                   24\n",
       "selection method                  25\n",
       "cell_type                         26\n",
       "store_cond                        26\n",
       "body_site                         27\n",
       "time point                        28\n",
       "birth_location                    28\n",
       "condition for cell growth         28\n",
       "molecular subtype                 29\n",
       "sample material processing        30\n",
       "Title                             30\n",
       "method                            30\n",
       "biological_replicate              31\n",
       "samp_collect_device               33\n",
       "submitter handle                  34\n",
       "genotype/background               34\n",
       "biospecimen repository            34\n",
       "purification                      35\n",
       "env biome                         36\n",
       "treatment protocol                38\n",
       "sample type                       39\n",
       "antibody target description       40\n",
       "source                            40\n",
       "growth protocol                   42\n",
       "isolation-source                  43\n",
       "cell sorting                      46\n",
       "                               ...  \n",
       "environment (material)           220\n",
       "sample_title                     228\n",
       "study name                       251\n",
       "geo_loc_name                     255\n",
       "project_name                     275\n",
       "sample_type                      293\n",
       "treatment                        301\n",
       "description                      302\n",
       "Environment (Feature)            312\n",
       "antibody targetdescription       321\n",
       "project name                     367\n",
       "env_matter                       387\n",
       "environment (feature)            423\n",
       "biomaterial_provider             474\n",
       "rel_to_oxygen                    481\n",
       "title                            506\n",
       "environment (biome)              601\n",
       "cell type                        666\n",
       "SUBMITTER_ID                     719\n",
       "strain                           772\n",
       "SCIENTIFIC_NAME                  817\n",
       "tissue                           872\n",
       "env_material                    1005\n",
       "env_biome                       1042\n",
       "depth                           1687\n",
       "source_name                     1786\n",
       "env_feature                     1840\n",
       "isolation_source                2310\n",
       "TITLE                           2966\n",
       "DESCRIPTION                    17231\n",
       "Length: 100, dtype: int64"
      ]
     },
     "execution_count": 370,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VC.sort_values().tail(n=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
